src/pudl/extract/eiaaeo.py

```diff
@@ -0,0 +1,453 @@
+    categories.
+    """
+
+    class EntityType(Enum):
+        """These are the three types of entities in AEO."""
+
+        ROOT = 1001
+        CATEGORY = 1002
+        SERIES = 1003
+
+    @dataclass
+    class CheckSpec:
+        """Encapsulate shared checks for the taxonomy structure."""
+
+        generation: str
+        typecheck: Callable[[int | str], bool]
+        in_degree: Callable[[int], bool]
+        out_degree: Callable[[int], bool]
+
+    def __init__(self, records: Iterable[str]):
+        """Load AEO JSON records into a graph datastructure.
+
+        Args:
+            records: the strings that contain an AEO report to parse.
+
+        """
+        categories, series = self.__load_records(records)
+        self.graph = self.__generate_graph(categories, series)
+        generations = self.__generation_invariants()
+        self.__cases = generations[1]
+        self.__sanitize_re = re.compile(r"\W+")
+
+    def __load_records(
+        self, records: Iterable[str]
+    ) -> tuple[dict[int, AEOCategory], dict[str, AEOSeries]]:
+        """Read AEO JSON blob into memory.
+
+        A record can be either a category or a series, so we parse those into
+        two separate mappings.
+        """
+        all_categories: dict[int, AEOCategory] = {}
+        all_series: dict[str, AEOSeries] = {}
+        for record in records:
+            if "category_id" in record:
+                category = AEOCategory.model_validate_json(record)
+                all_categories[category.category_id] = category
+            elif "series_id" in record:
+                series = AEOSeries.model_validate_json(record)
+                all_series[series.series_id] = series
+            else:
+                raise ValueError(f"Line had neither series nor category ID: {record}")
+        return all_categories, all_series
+
+    def __generate_graph(
+        self, categories: dict[int, AEOCategory], series: dict[str, AEOSeries]
+    ) -> nx.DiGraph:
+        """Stitch categories and series together into a DAG."""
+
+        def get_all_edges(category: AEOCategory) -> Iterable[tuple[int, str | int]]:
+            # A category can have incoming edges from its parents or outgoing
+            # edges to its child series.
+            edges: list[tuple[int, str | int]] = [
+                (category.parent_category_id, category.category_id)
+            ]
+            if category.childseries:
+                edges += [
+                    (category.category_id, child_series_id)
+                    for child_series_id in category.childseries
+                ]
+            return edges
+
+        edges = itertools.chain.from_iterable(
+            get_all_edges(category) for category in categories.values()
+        )
+        graph = nx.DiGraph(incoming_graph_data=edges)
+
+        nx.set_node_attributes(graph, categories | series)
+        return graph
+
+    def __generation_invariants(self) -> list:  # noqa: C901
```

Comment:
I feel like there are a bunch of error checking patterns encapsulated in this method that we could learn from and deploy elsewhere.